### PR TITLE
cpu/qn908x: Implement blocking I2C support

### PR DESCRIPTION
**Pull request by iosabi**
### Contribution description

This initial I2C support allows to use the I2C bus in controller mode to
interact with multiple peripherals in blocking mode. The CPU will
perform a busy wait when transferring data over I2C.

This patch also adds the driver for the MMA8652FC and the I2C
configuration to the qn9080dk board.

### Testing procedure

Ran `i2c_scan` and found all devices connected to the bus, correctly marking the free addresses with `-`.

Used this driver to run `si1133` driver tests.

### Issues/PRs references

This is part of FR #13852 .

### CODE REVIEW

File: cpu/qn908x/doc.txt [hunk 1]
```diff
@@ -85,13 +114,13 @@ other pin can be set to GPIO_UNDEF.
...
-            .tx_pin       = GPIO_PIN(PORT_A, 16), /* or 4 */
+            .rx_pin       = GPIO_PIN(PORT_A, 17),  // or 5
+            .tx_pin       = GPIO_PIN(PORT_A, 16),  // or 4
         },
         {
             .dev          = USART1,
-            .rx_pin       = GPIO_PIN(PORT_A, 9), /* or 13 */
-            .tx_pin       = GPIO_PIN(PORT_A, 8), /* or 12 */
+            .rx_pin       = GPIO_PIN(PORT_A, 9),  // or 13
+            .tx_pin       = GPIO_PIN(PORT_A, 8),  // or 12
```
Comment (benpicco) [inline cpu/qn908x/doc.txt:123]: let's keep the C style comments :wink:

File: cpu/qn908x/include/periph_cpu.h [hunk 2]
```diff
@@ -149,6 +149,51 @@ enum {
...
+/** @} */
+
+/**
+ * @brief   I2C configuration options
+ */
+typedef struct {
+    I2C_Type *dev;          /**< hardware device */
+    gpio_t pin_scl;         /**< SCL pin */
+    gpio_t pin_sda;         /**< SDA pin */
+    uint32_t speed;         /**< bus speed in kbit/s */
```
Comment (benpicco) [inline cpu/qn908x/include/periph_cpu.h:176]: might also use `uint16_t` if 400 is the largest value.
`uint32_t` makes it look like `speed` was in bit/s.

File: cpu/qn908x/doc.txt [hunk 3]
```diff
@@ -31,6 +31,35 @@ The GPIO driver uses the @ref GPIO_PIN(port, pin) macro to declare pins.
...
+        {
+            .dev          = I2C0,
+            .pin_scl      = GPIO_PIN(PORT_A, 6),  // or 8, 12, 20
+            .pin_sda      = GPIO_PIN(PORT_A, 7),  // or 9, 13, 21
+            .speed        = I2C_SPEED_FAST,
+        },
+        {
+            .dev          = I2C1,
+            .pin_scl      = GPIO_PIN(PORT_A, 2),  // or 5, 23, 27
+            .pin_sda      = GPIO_PIN(PORT_A, 3),  // or 4, 22, 26
```
Comment (benpicco) [inline cpu/qn908x/doc.txt:56]: You mean A4, A22, A26?

File: cpu/qn908x/periph/i2c.c [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+    assert(dev < I2C_NUMOF);
+    const i2c_conf_t *conf = &i2c_config[dev];
+    I2C_Type *const i2c_dev = conf->dev;
+
+    int flexcomm_num = flexcomm_init((FLEXCOMM_Type *)i2c_dev, FLEXCOMM_ID_I2C);
+    assert(flexcomm_num >= 0);
+
+    if (flexcomm_num == 1) {
+        /* Disable the FLECOMM1 MULT1/DIV1 divisor. */
+        CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0u);
```
Comment (benpicco) [inline cpu/qn908x/periph/i2c.c:114]: Why is FLECOMM1 special?

File: boards/qn9080dk/Makefile.dep [hunk 5]
```diff
@@ -1,3 +1,4 @@
 ifneq (,$(filter saul_default,$(USEMODULE)))
   USEMODULE += saul_gpio
+  USEMODULE += mma8x5x
```
Comment (benpicco) [inline boards/qn9080dk/Makefile.dep:3]: let's keep that in a separate commit

**Review by benpicco — COMMENTED** [review]
Looking good!
One nitpick though: I'd split that into three commits

 - one commit implementing the i2c driver for `cpu/qn908x`
 - one commit adding the board configuration for `boards/qn9080dk`
 - one commit enabling `mma8x5x` on `qn9080dk`

**Comment by iosabi** [comment]
> Looking good!
> One nitpick though: I'd split that into three commits
> 
> * one commit implementing the i2c driver for `cpu/qn908x`
> * one commit adding the board configuration for `boards/qn9080dk`
> * one commit enabling `mma8x5x` on `qn9080dk`

Do you want 3 commits in the same PR? or three different PR? (the idea is to have a board that builds this). There's some guidance in the website that says 1 commit per PR.

File: cpu/qn908x/doc.txt (hunk 1 above)
Comment (iosabi) [reply cpu/qn908x/doc.txt:123]: I'd like to, but the doc.txt is actually a giant C comment that starts with `/**` at the top of the file and this comment here was closing it, dropping everything after this comment from doxygen.

**Comment by benpicco** [comment]
3 commits in the same PR

> There's some guidance in the website that says 1 commit per PR.

Huh, that sounds wrong. Where is that?

**Comment by iosabi** [comment]
> 3 commits in the same PR
> 
> > There's some guidance in the website that says 1 commit per PR.
> 
> Huh, that sounds wrong. Where is that?

https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#git

Now that you mention it I guess I misread that as 1 commit per PR, but it probably means one commit per "logical change" or "conceptually independent change". I think that's where I had the impression of making 1 commit.

File: cpu/qn908x/periph/i2c.c (hunk 4 above)
Comment (iosabi) [reply cpu/qn908x/periph/i2c.c:114]: FLEXCOMM0 and 1 support UART and they both have a "fractional" clock divider set for both 0 and 1 in the FC_FRG register. This is different from all the other frequency dividers because instead of taking a clock signal and divide by N which is normally done with a simple counter circuit up to N, this "fractional" one allows you to divide your clock frequency by (1+N/256) where N is between 0 and 255, so you divide by something between 1 and 2. My personal guess is that this is useful in the UART case to run the UART at 9600 bps when your source clock is running at 32 or 32.768 KHz. With only a divisor like you have in the AHB and APB buses your "standard" UART in that case is quite off (8192 or 10922). I think this has some application when going into power down mode but leaving the UART enabled, I haven't tried that.

Anyway, it looks from the manual that this fractional is before the "Flexcomm clock", so this is active even if you use the FLEXCOMM1 for I2C. Your question got me thinking so I hooked up the oscilloscope to check and indeed the I2C clock I see is affected by this. Given the typical frequencies involved in I2C and the fact that it is synchronous so nobody cares about the exact value, I opted for disabling it. This code was setting N to 0, so we divide by (1 + 0/256) which is 1.

FLEXCOMM2 ( == I2C1) doesn't have the fractional divisor.

File: cpu/qn908x/include/periph_cpu.h (hunk 2 above)
Comment (iosabi) [reply cpu/qn908x/include/periph_cpu.h:176]: I made it uint16_t.

File: cpu/qn908x/doc.txt (hunk 3 above)
Comment (iosabi) [reply cpu/qn908x/doc.txt:56]: Done.

File: boards/qn9080dk/Makefile.dep (hunk 5 above)
Comment (iosabi) [reply boards/qn9080dk/Makefile.dep:3]: Done.

File: cpu/qn908x/periph/i2c.c (hunk 4 above)
Comment (benpicco) [reply cpu/qn908x/periph/i2c.c:125]: Sounds a lot like Atmel's SERCOMs.

But shouldn't this then be 

```C
if (flexcomm_num == 0 || flexcomm_num == 1) {
```

File: cpu/qn908x/periph/i2c.c [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    if (flexcomm_num == 1) {
+        /* Disable the FLECOMM1 MULT1/DIV1 divisor. FLEXCOMM2 doesn't have a
+         * fractional divisor. This divides the clock by (1 + N / 256) where
+         * N is set to 0 here, so the fractional divisor only divides by one. */
+        CLOCK_SetClkDiv(kCLOCK_DivFrg1, 0u);
+    }
+
+    /* Enable controller mode, no timeout, no monitor, no clock stretching. */
+    i2c_dev->CFG = I2C_CFG_MSTEN_MASK;
+    _i2c_controller_set_speed(i2c_dev, conf->speed * 1000);
```
Comment (benpicco) [inline cpu/qn908x/periph/i2c.c:130]: Makes me wonder: why not have `i2c_speed_t` as bit/s 
That would be consistent with other architectures that allow just any value here.

File: cpu/qn908x/periph/i2c.c (hunk 4 above)
Comment (iosabi) [reply cpu/qn908x/periph/i2c.c:125]: In theory yes, and also use `kCLOCK_DivFrg0` instead of 1 for `flexcomm_num == 0`, but the FLEXCOMM0 doesn't support I2C so we would never be here already.

File: cpu/qn908x/periph/i2c.c (hunk 6 above)
Comment (iosabi) [reply cpu/qn908x/periph/i2c.c:130]: Sure.

**Review by benpicco — APPROVED** [review]
Looks good to me!

File: cpu/qn908x/periph/i2c.c (hunk 4 above)
Comment (benpicco) [reply cpu/qn908x/periph/i2c.c:125]: ah turns out those FLEXCOMMs are not as flexible as I thought!
I guess the allocation might change with new members of this chip family, so it would probably good to leave a comment explaining that only FLEXCOM1 and FLEXCOM2 can be I2C.

Comment (iosabi) [reply cpu/qn908x/periph/i2c.c:125]: Haha, marketing got you. All the FLEXCOMM blocks have a bit mask at the same offset (0xFF8, PSELID register) that says what functions (I2C, UART, SPI, ...) the given flexcomm supports, and a hardware revision number in case you need to rely on it to know which bugs does the FLEXCOMM block have. I think the idea is more about writing "Flexcomm drivers" that you can share across devices of revisions of the devices and check at runtime that the hardware you are talking to supports your driver, or maybe even discover the capabilities at runtime. For example, the nxp LPC546xx have up to 10 FLEXCOMM and all of them support UART, SPI and I2C, but only two of those support I2S as well. A cursory look at the LPC546xx datasheet shows that the list of registers for those FLEXCOMM are very similar to the QN9080, expect that the QN9080 datasheet shows "reserved" for the bits used for the I2S in that LPC. So there's some potential for reusing these.

That said, the "fractional" divider is not part of the FLEXCOMM block itself, the register this function is changing (`SYSCON->FC_FRG`) is part of the SYSCON block, not the I2C/FLEXCOMM one. In the USART mode I select from `uart.c` the "fractional" divider in SYSCON and the BRG / OSR values in the USARTn/FLEXCOMMn block; so that part is shared. In the LPC546xx example there is apparently only one fractional divider shared among all 10 FLEXCOMM blocks. It kinda makes sense since whether you use the fractional divider depends on whether the AHB is running at 32 KHz or not.

Dividing by 1 is the reset deafult value (MCU reset, not FLEXCOMM reset because the value is in SYSCON). So if you limit this to call `i2c_init()` you are fine even without this call, but if for some reason you are switching between I2C and UART it is possible that the fractional divider was set to something else. What I can do instead is to make `flexcomm_init()` set that value to 0 for the flexcomms that have one (hard-coded to 0 and 1 in this MCU, since this is not really a flexcomm feature), and remove that call from the i2c.c driver. The uart.c driver will set the fractional divider on initialization anyway.
